src/client/components/vision2/camera/greenhorizon.ts

```diff
@@ -0,0 +1,63 @@
+import { IComputedValue } from 'mobx'
+import { computed } from 'mobx'
+import * as THREE from 'three'
+
+import { range } from '../../../../shared/base/range'
+import { Matrix4 } from '../../../math/matrix4'
+import { Vector3 } from '../../../math/vector3'
+import { Vector4 } from '../../../math/vector4'
+import { group } from '../../three/builders'
+import { Canvas } from '../../three/three'
+import { GreenHorizon } from '../../vision/camera/model'
+
+import { Lens } from './model'
+import { WorldLine } from './world_line'
+import { coneSegment } from './world_line'
+
+export class GreenHorizonViewModel {
+  constructor(
+    private readonly model: () => GreenHorizon,
+    private readonly imageHcw: () => Matrix4,
+    private readonly worldLine: WorldLine,
+  ) {
+  }
+
+  static of(
+    canvas: Canvas,
+    model: () => GreenHorizon,
+    imageHcw: () => Matrix4,
+    lens: () => Lens,
+  ): GreenHorizonViewModel {
+    return new GreenHorizonViewModel(model, imageHcw, WorldLine.of(canvas, lens))
+  }
+
+  readonly greenhorizon = group(() => ({
+    children: range(this.model().horizon.length - 1).map(index => {
+      return this.worldLine.planeSegment(this.segment(index))
+    }),
+  }))
+
+  private readonly segment = coneSegment((index: number) => ({
+    start: this.rays[index],
+    end: this.rays[index + 1],
+    color: new Vector4(0, 0.8, 0, 0.8),
+    lineWidth: 10,
+  }))
+
+  @computed
+  private get rays() {
+    const { horizon, Hcw: oldHcw } = this.model()
+    const newHcw = this.imageHcw()
+    const oldHwc = Matrix4.fromThree(new THREE.Matrix4().getInverse(oldHcw.toThree()))
```

Comment:
I think old and new aren't the best names here. It's more like, `imageHcw` or `projectionHcw` rather than new and `measurementHcw` or `rayHcw`. Or some other synonym for those. There's no reason that the GreenHorizon Hcw isn't newer than the image one in terms of time. Also i'm not 100% sure on how to fix it but sticking things on the Hcw etc variables always looks weird